File: app/lib/features/permissions/domain/entities/permission_prompt_state_entity.dart
import 'package:dart_mappable/dart_mappable.dart';

part 'permission_prompt_state_entity.mapper.dart';

/// Preferência de produto para uma permissão de sistema (evita repetir onboarding).
@MappableClass()
class PermissionPromptRecordEntity with PermissionPromptRecordEntityMappable {
  const PermissionPromptRecordEntity({
    this.seen = false,
    this.skipped = false,
    this.completedAt,
  });

  final bool seen;
  final bool skipped;
  final DateTime? completedAt;
}

/// Agregado das telas de permissão (notificação, localização, biometria).
@MappableClass()
class PermissionPromptStateEntity with PermissionPromptStateEntityMappable {
  const PermissionPromptStateEntity({
    this.notification = const PermissionPromptRecordEntity(),
    this.location = const PermissionPromptRecordEntity(),
    this.biometric = const PermissionPromptRecordEntity(),
  });

  final PermissionPromptRecordEntity notification;
  final PermissionPromptRecordEntity location;
  final PermissionPromptRecordEntity biometric;
}

File: backend/functions/src/features/authentication/domain/entities/permission_prompt_state_entity.ts
import {z} from "zod";

/**
 * Preferência de produto para permissão de sistema (evita repetir onboarding).
 */
export const permissionPromptRecordEntitySchema = z.object({
  seen: z.boolean().default(false),
  skipped: z.boolean().default(false),
  completedAt: z.coerce.date().optional(),
});

export type PermissionPromptRecordEntity = z.infer<
  typeof permissionPromptRecordEntitySchema
>;

/** Agregado das telas de permissão (notificação, localização, biometria). */
export const permissionPromptStateEntitySchema = z.object({
  notification: permissionPromptRecordEntitySchema.default({
    seen: false,
    skipped: false,
  }),
  location: permissionPromptRecordEntitySchema.default({
    seen: false,
    skipped: false,
  }),
  biometric: permissionPromptRecordEntitySchema.default({
    seen: false,
    skipped: false,
  }),
});

export type PermissionPromptStateEntity = z.infer<
  typeof permissionPromptStateEntitySchema
>;
